service/cachemdw/middleware.go

```diff
@@ -0,0 +1,109 @@
+package cachemdw
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+
+	"github.com/kava-labs/kava-proxy-service/decode"
+)
+
+type cacheContextKey string
+
+const (
+	CachedContextKey cacheContextKey = "X-KAVA-PROXY-CACHED"
+
+	CacheHeaderKey       = "X-Cache"
+	CacheHitHeaderValue  = "HIT"
+	CacheMissHeaderValue = "MISS"
+)
+
+func (c *ServiceCache) Middleware(
+	next http.Handler,
+) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		uncachedContext := context.WithValue(r.Context(), CachedContextKey, false)
+
+		req := r.Context().Value(c.decodedRequestContextKey)
+		decodedReq, ok := (req).(*decode.EVMRPCRequestEnvelope)
+		if !ok {
+			next.ServeHTTP(w, r.WithContext(uncachedContext))
+			return
+		}
+
+		cachedQueryResponse, err := c.GetCachedQueryResponse(r.Context(), decodedReq, r.Host)
+		if err == nil {
+			c.respond(next, w, r, cachedQueryResponse)
+			return
+		}
+
+		cacheable := IsCacheable(r.Context(), c.evmClient, decodedReq)
+		if !cacheable {
+			next.ServeHTTP(w, r.WithContext(uncachedContext))
+			return
+		}
+
+		c.respondAndCache(uncachedContext, next, w, r, decodedReq)
+	}
+}
+
+func (c *ServiceCache) respond(
+	next http.Handler,
+	w http.ResponseWriter,
+	r *http.Request,
+	cachedQueryResponse []byte,
+) {
+	w.Header().Add(CacheHeaderKey, CacheHitHeaderValue)
+	w.Header().Add("Content-Type", "application/json")
+	_, err := w.Write(cachedQueryResponse)
+	if err != nil {
+		c.Logger.Error().Msg(fmt.Sprintf("can't write cached response: %v", err))
+	}
+
+	cachedContext := context.WithValue(r.Context(), CachedContextKey, true)
+	next.ServeHTTP(w, r.WithContext(cachedContext))
+}
+
+func (c *ServiceCache) respondAndCache(
+	uncachedContext context.Context,
+	next http.Handler,
+	w http.ResponseWriter,
+	r *http.Request,
+	decodedReq *decode.EVMRPCRequestEnvelope,
+) {
+	w.Header().Add(CacheHeaderKey, CacheMissHeaderValue)
+
+	recorder := httptest.NewRecorder()
+	next.ServeHTTP(recorder, r.WithContext(uncachedContext))
```

Comment:
nope, use this https://github.com/Kava-Labs/kava-proxy-service/blob/main/service/middleware.go#L42
not using the test package for production code :face-palm:

Reply:
done